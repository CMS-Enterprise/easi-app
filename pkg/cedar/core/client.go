package cedarcore

import (
	"context"
	"net/http"
	"time"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
)

const (
	cedarCoreCacheDurationDefault = time.Hour * 6
)

type loggingTransport struct {
	logger *zap.Logger
}

func (t *loggingTransport) RoundTrip(r *http.Request) (*http.Response, error) {
	start := time.Now().UnixMilli()
	resp, err := http.DefaultTransport.RoundTrip(r)
	end := time.Now().UnixMilli()

	// Start a status code of 0, in case the request fails (and we get a nil resp)
	status := 0
	if resp != nil {
		status = resp.StatusCode
	}

	t.logger.Info(
		"Call to CEDAR core",
		zap.String("service", "cedarcore"),
		zap.Bool("cacheEnabled", false),
		zap.String("method", r.Method),
		zap.Int("status", status),
		zap.String("path", r.URL.Path),
		zap.String("queryParams", r.URL.RawQuery),
		zap.Int64("timeMS", end-start),
	)

	return resp, err
}

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, cedarAPIVersion string, cacheRefreshTime time.Duration, mockEnabled bool) *Client {
	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.Client{
		Transport: &loggingTransport{
			logger: appcontext.ZLogger(ctx),
		},
	}

	basePath := "/gateway/CEDAR Core API/" + cedarAPIVersion
	client := &Client{
		mockEnabled: mockEnabled,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				basePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    &hc,
		cache: c,
	}

	if cacheRefreshTime <= 0 {
		appcontext.ZLogger(ctx).Error(
			"CEDAR cacheRefreshTime is not a valid value, falling back to default",
			zap.Duration("cacheRefreshTime", cacheRefreshTime),
			zap.Duration("defaultDuration", cedarCoreCacheDurationDefault),
		)
		cacheRefreshTime = cedarCoreCacheDurationDefault
	}
	client.startCacheRefresh(ctx, cacheRefreshTime, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	mockEnabled bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARCoreAPI
	hc          *http.Client
	cache       *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime.
// startCacheRefresh returns no errors, and only logs when something goes wrong.
// Upon being called, startCacheRefresh will populate the cache once immediately, then again at an interval specificed by cacheRefreshTime.
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				// This is a Warn() instead of Error() as it happens somewhat frequently, and doesn't necessarily warrant immediate attention
				appcontext.ZLogger(ctx).Warn("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
