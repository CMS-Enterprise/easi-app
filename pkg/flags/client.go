package flags

import (
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v4"

	"github.com/cmsgov/easi-app/pkg/appconfig"
)

// Config has all the parts for creating a new LD Client
type Config struct {
	Source  appconfig.FlagSourceOption
	Key     string
	Timeout time.Duration
	Offline bool
}

// FlagValues is the struct for our keys and values from LD
type FlagValues map[string]interface{}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	ldConfig := ld.Config{Offline: config.Offline}
	return ld.MakeCustomClient(config.Key, ldConfig, config.Timeout)
}
