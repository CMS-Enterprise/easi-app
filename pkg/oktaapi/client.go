// package oktaapi

// import (
// 	"context"

// 	"github.com/cmsgov/easi-app/pkg/models"
// )

// // Client defines methods which any Okta API Client should implement
// type Client interface {
// 	FetchUserInfo(context.Context, string) (*models.UserInfo, error)
// 	SearchByName(context.Context, string) ([]*models.UserInfo, error)
// }
