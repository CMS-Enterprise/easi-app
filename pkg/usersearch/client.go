package usersearch

import (
	"context"

	"github.com/cmsgov/easi-app/pkg/models"
)

// Client is a generic interface for userSearch API
type Client interface {
	FetchUserInfo(context.Context, string) (*models.UserInfo, error)
	FetchUserInfos(context.Context, []string) ([]*models.UserInfo, error)
	SearchCommonNameContains(context.Context, string) ([]*models.UserInfo, error)
}
