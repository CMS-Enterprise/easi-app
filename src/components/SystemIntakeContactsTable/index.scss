@use 'viewports' as *;

// Fix for tooltip content cutting off when table is scrollable
.system-intake-contacts-table.usa-table-container--scrollable {
  min-height: 140px;

  .usa-tooltip__body {
    white-space: pre-wrap;
    min-width: 240px;
  }
  
  @media (min-width: $desktop) {
    overflow-y: visible !important;
    min-height: auto;
  }
}
