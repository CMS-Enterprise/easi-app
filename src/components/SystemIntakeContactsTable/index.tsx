import React, { useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Column, useSortBy, useTable } from 'react-table';
import {
  Button,
  ButtonGroup,
  Icon,
  ModalFooter,
  ModalHeading,
  Table,
  Tooltip
} from '@trussworks/react-uswds';
import classNames from 'classnames';
import {
  SystemIntakeContactFragment,
  SystemIntakeContactRole
} from 'gql/generated/graphql';

import Modal from 'components/Modal';
import Spinner from 'components/Spinner';
import cmsComponentsMap from 'constants/cmsComponentsMap';
import { getColumnSortStatus, getHeaderSortIcon } from 'utils/tableSort';

import './index.scss';

type SystemIntakeContactsTableProps = {
  contacts: SystemIntakeContactFragment[] | undefined;
  handleEditContact?: (contact: SystemIntakeContactFragment) => void;
  removeContact?: (id: string) => void;
  /** If true, a loading spinner and text will render in place of results */
  loading?: boolean;
  className?: string;
};

/**
 * Table component to display system intake contacts
 *
 * Renders actions column and button(s) if handleEditContact and/or removeContact is provided
 */
const SystemIntakeContactsTable = ({
  contacts,
  handleEditContact,
  removeContact,
  loading,
  className
}: SystemIntakeContactsTableProps) => {
  const { t } = useTranslation('intake');

  const [contactIdToRemove, setContactIdToRemove] = useState<string | null>(
    null
  );

  /** Returns true if either `handleEditContact` or `removeContact` is provided */
  const hasActionsColumn = useMemo(
    () => !!handleEditContact || !!removeContact,
    [handleEditContact, removeContact]
  );

  const columns = useMemo<Column<SystemIntakeContactFragment>[]>(() => {
    return [
      {
        // createdAt column is hidden and only used for sorting purposes
        accessor: 'createdAt',
        id: 'createdAt'
      },
      {
        Header: () => (
          <span className="display-block margin-left-4 padding-left-05">
            {t('general:name')}
          </span>
        ),
        accessor: (row: SystemIntakeContactFragment) =>
          row.userAccount.commonName,
        id: 'commonName',
        width: handleEditContact ? 320 : 'auto',
        Cell: ({ row }: { row: { original: SystemIntakeContactFragment } }) => (
          <div className="display-flex flex-align-center">
            {row.original.isRequester && (
              <Tooltip
                position="top"
                label={t('contactDetails.additionalContacts.requesterTooltip')}
                className="padding-0 margin-0 margin-top-1 bg-transparent outline-0"
              >
                <Icon.ContactPage
                  className="text-primary"
                  size={3}
                  aria-label={t(
                    'contactDetails.additionalContacts.primaryRequester'
                  )}
                />
              </Tooltip>
            )}
            <div
              className={classNames(
                row.original.isRequester
                  ? 'margin-left-105'
                  : 'margin-left-4 padding-left-05'
              )}
            >
              <p data-testid="commonName">
                {row.original.userAccount.commonName}
              </p>
              <p data-testid="email">{row.original.userAccount.email}</p>
            </div>
          </div>
        )
      },
      {
        Header: t('fields.component'),
        accessor: 'component',
        id: 'component',
        width: handleEditContact ? 150 : 200,
        Cell: ({
          value
        }: {
          value: SystemIntakeContactFragment['component'];
        }) => {
          if (!value) {
            return (
              <span className="text-base-dark text-italic">
                {t('general:noneSpecified')}
              </span>
            );
          }

          // Get component data using enum value
          const component = cmsComponentsMap[value];

          // Display acronym if available, otherwise display component name
          return <>{component.acronym || t(component.name)}</>;
        }
      },
      {
        Header: t('fields.roles'),
        accessor: 'roles',
        id: 'roles',
        width: 'auto',
        Cell: ({ value: roles }: { value: SystemIntakeContactRole[] }) => {
          if (roles.length === 0) {
            return (
              <span className="text-base-dark text-italic">
                {t('general:noneSpecified')}
              </span>
            );
          }

          // Return list of translated enum values
          return (
            <>
              {roles
                .map(role =>
                  t(`contactDetails.systemIntakeContactRoles.${role}`)
                )
                .join(', ')}
            </>
          );
        }
      },
      {
        Header: t('general:actions'),
        id: 'actions',
        accessor: (row: SystemIntakeContactFragment, index) => {
          return (
            <div className="display-flex">
              {handleEditContact && (
                <Button
                  type="button"
                  className="margin-top-0 margin-right-2"
                  onClick={() => handleEditContact(row)}
                  data-testid={`editContact-${index}`}
                  unstyled
                >
                  {t('general:edit')}
                </Button>
              )}
              {removeContact && (
                <Button
                  type="button"
                  className={classNames(
                    'margin-top-0',
                    row.isRequester ? 'text-disabled' : 'text-error'
                  )}
                  unstyled
                  onClick={() => setContactIdToRemove(row.id)}
                  data-testid={`removeContact-${index}`}
                  disabled={row.isRequester}
                >
                  {t('general:remove')}
                </Button>
              )}
            </div>
          );
        },
        width: 140
      }
    ];
  }, [t, handleEditContact, removeContact]);

  const table = useTable(
    {
      columns,
      data: contacts || [],
      autoResetSortBy: false,
      autoResetPage: true,
      initialState: {
        hiddenColumns: useMemo(
          () => [
            'createdAt',
            // Hide actions column if `hasActionsColumn` is false
            ...(hasActionsColumn ? [] : ['actions'])
          ],
          [hasActionsColumn]
        ),
        sortBy: useMemo(() => [{ id: 'createdAt', desc: false }], [])
      }
    },
    useSortBy
  );

  const { getTableBodyProps, getTableProps, headerGroups, prepareRow, rows } =
    table;

  rows.map(row => prepareRow(row));

  return (
    <>
      <div
        className={classNames(
          'system-intake-contacts-table usa-table-container--scrollable',
          className
        )}
      >
        <Table bordered={false} fullWidth {...getTableProps()}>
          <thead>
            {headerGroups.map(headerGroup => {
              const { key, ...headerGroupProps } =
                headerGroup.getHeaderGroupProps();

              return (
                <tr {...headerGroupProps} key={key}>
                  {headerGroup.headers.map(column => {
                    const { key: headerKey, ...headerProps } =
                      column.getHeaderProps();

                    if (column.id === 'actions') {
                      return (
                        <th
                          {...headerProps}
                          key={headerKey}
                          className="border-bottom-2px"
                          style={{
                            width: column.width
                          }}
                        >
                          {column.render('Header')}
                        </th>
                      );
                    }

                    return (
                      <th
                        {...headerProps}
                        key={headerKey}
                        aria-sort={getColumnSortStatus(column)}
                        scope="col"
                        className="border-bottom-2px"
                        style={{
                          width: column.width
                        }}
                      >
                        <Button
                          type="button"
                          className="width-full flex-justify margin-y-0"
                          unstyled
                          {...column.getSortByToggleProps()}
                        >
                          {column.render('Header')}
                          {getHeaderSortIcon(column)}
                        </Button>
                      </th>
                    );
                  })}
                </tr>
              );
            })}
          </thead>
          <tbody {...getTableBodyProps()}>
            {rows.length > 0 ? (
              rows.map(row => {
                const { key: rowKey, ...rowProps } = row.getRowProps();

                const { id } = row.original;

                return (
                  <tr
                    {...rowProps}
                    key={rowKey}
                    data-testid="contact-row"
                    aria-label={`contact-${id}`}
                  >
                    {row.cells.map(cell => {
                      const { key: cellKey, ...cellProps } =
                        cell.getCellProps();

                      return (
                        <td {...cellProps} key={cellKey}>
                          {cell.render('Cell')}
                        </td>
                      );
                    })}
                  </tr>
                );
              })
            ) : (
              <tr>
                <td colSpan={columns.length} className="padding-left-4">
                  <span className="margin-left-05 display-flex flex-align-center text-italic">
                    {loading ? (
                      <>
                        <Spinner size="small" className="margin-right-1" />
                        {t('contactDetails.loadingContacts')}
                      </>
                    ) : (
                      t('contactDetails.noContacts')
                    )}
                  </span>
                </td>
              </tr>
            )}
          </tbody>
        </Table>
      </div>

      {/* Remove contact modal */}
      {removeContact && (
        <Modal
          isOpen={!!contactIdToRemove}
          closeModal={() => setContactIdToRemove(null)}
          className="font-body-md"
        >
          <ModalHeading>
            {t('contactDetails.additionalContacts.removeModal.heading')}
          </ModalHeading>
          <p>
            {t('contactDetails.additionalContacts.removeModal.description')}
          </p>

          <ModalFooter>
            <ButtonGroup>
              <Button
                type="button"
                className="margin-right-2 bg-error"
                onClick={() => {
                  if (contactIdToRemove) {
                    removeContact(contactIdToRemove);
                  }

                  setContactIdToRemove(null);
                }}
              >
                {t('contactDetails.additionalContacts.removeModal.submit')}
              </Button>

              <Button
                type="button"
                onClick={() => setContactIdToRemove(null)}
                unstyled
              >
                {t('general:cancel')}
              </Button>
            </ButtonGroup>
          </ModalFooter>
        </Modal>
      )}
    </>
  );
};

export default SystemIntakeContactsTable;
