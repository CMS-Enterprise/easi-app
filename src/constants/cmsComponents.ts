import { SystemIntakeContactComponent } from 'gql/generated/graphql';

type ComponentsMap = Record<
  SystemIntakeContactComponent,
  {
    /** If true, component is only found in legacy data and cannot be selected in the UI */
    legacy: boolean;
    name: string;
    acronym?: string;
  }
>;

/** Map of CMS components to their name and acronym */
const cmsComponents: ComponentsMap = {
  CENTER_FOR_CLINICAL_STANDARDS_AND_QUALITY_CCSQ: {
    acronym: 'CCSQ',
    name: 'Center for Clinical Standards and Quality',
    legacy: false
  },
  CENTER_FOR_CONSUMER_INFORMATION_AND_INSURANCE_OVERSIGHT_CCIIO: {
    acronym: 'CCIIO',
    name: 'Center for Consumer Information and Insurance Oversight',
    legacy: false
  },
  CENTER_FOR_MEDICARE_CM: {
    acronym: 'CM',
    name: 'Center for Medicare',
    legacy: false
  },
  CENTER_FOR_MEDICAID_AND_CHIP_SERVICES_CMCS: {
    acronym: 'CMCS',
    name: 'Center for Medicaid and CHIP Services',
    legacy: false
  },
  CENTER_FOR_MEDICARE_AND_MEDICAID_INNOVATION_CMMI: {
    acronym: 'CMMI',
    name: 'Center for Medicare and Medicaid Innovation',
    legacy: false
  },
  CENTER_FOR_PROGRAM_INTEGRITY_CPI: {
    acronym: 'CPI',
    name: 'Center for Program Integrity',
    legacy: false
  },
  CMS_WIDE: { acronym: 'CMS', name: 'CMS Wide', legacy: false },
  EMERGENCY_PREPAREDNESS_AND_RESPONSE_OPERATIONS_EPRO: {
    acronym: 'EPRO',
    name: 'Emergency Preparedness and Response Operations',
    legacy: false
  },
  FEDERAL_COORDINATED_HEALTH_CARE_OFFICE: {
    acronym: 'FCHCO',
    name: 'Federal Coordinated Health Care Office',
    legacy: false
  },
  OFFICE_OF_ACQUISITION_AND_GRANTS_MANAGEMENT_OAGM: {
    acronym: 'OAGM',
    name: 'Office of Acquisition and Grants Management',
    legacy: false
  },
  OFFICE_OF_HEALTHCARE_EXPERIENCE_AND_INTEROPERABILITY: {
    acronym: 'OHEI',
    name: 'Office of Healthcare Experience and Interoperability',
    legacy: false
  },
  OFFICE_OF_COMMUNICATIONS_OC: {
    acronym: 'OC',
    name: 'Office of Communications',
    legacy: false
  },
  OFFICE_OF_ENTERPRISE_DATA_AND_ANALYTICS_OEDA: {
    acronym: 'OEDA',
    name: 'Office of Enterprise Data and Analytics',
    legacy: false
  },
  OFFICE_OF_EQUAL_OPPORTUNITY_AND_CIVIL_RIGHTS: {
    acronym: 'OEOCR',
    name: 'Office of Equal Opportunity and Civil Rights',
    legacy: false
  },
  OFFICE_OF_FINANCIAL_MANAGEMENT_OFM: {
    acronym: 'OFM',
    name: 'Office of Financial Management',
    legacy: false
  },
  OFFICE_OF_HUMAN_CAPITAL: {
    acronym: 'OHC',
    name: 'Office of Human Capital',
    legacy: false
  },
  OFFICE_OF_INFORMATION_TECHNOLOGY_OIT: {
    acronym: 'OIT',
    name: 'Office of Information Technology',
    legacy: false
  },
  OFFICE_OF_LEGISLATION: {
    acronym: 'OL',
    name: 'Office of Legislation',
    legacy: false
  },
  OFFICE_OF_MINORITY_HEALTH_OMH: {
    acronym: 'OMH',
    name: 'Office of Minority Health',
    legacy: false
  },
  OFFICE_OF_PROGRAM_OPERATIONS_AND_LOCAL_ENGAGEMENT_OPOLE: {
    acronym: 'OPOLE',
    name: 'Office of Program Operations and Local Engagement',
    legacy: false
  },
  OFFICE_OF_SECURITY_FACILITIES_AND_LOGISTICS_OPERATIONS_OSFLO: {
    acronym: 'OSFLO',
    name: 'Office of Security, Facilities, and Logistics Operations',
    legacy: false
  },
  OFFICE_OF_STRATEGIC_OPERATIONS_AND_REGULATORY_AFFAIRS_OSORA: {
    acronym: 'OSORA',
    name: 'Office of Strategic Operations and Regulatory Affairs',
    legacy: false
  },
  OFFICE_OF_STRATEGY_PERFORMANCE_AND_RESULTS_OSPR: {
    acronym: 'OSPR',
    name: 'Office of Strategy, Performance, and Results',
    legacy: false
  },
  OFFICE_OF_THE_ACTUARY_OACT: {
    acronym: 'OACT',
    name: 'Office of the Actuary',
    legacy: false
  },
  OFFICES_OF_HEARINGS_AND_INQUIRIES: {
    acronym: 'OHI',
    name: 'Offices of Hearings and Inquiries',
    legacy: false
  },
  OFFICE_OF_THE_ADMINISTRATOR: {
    acronym: 'OA',
    name: 'Office Of The Administrator',
    legacy: false
  },
  CONSORTIUM_FOR_MEDICAID_AND_CHILDRENS_HEALTH: {
    name: "Consortium For Medicaid And Children's Health",
    legacy: true
  },
  CONSORTIUM_FOR_MEDICARE_HEALTH_PLANS_OPERATIONS: {
    name: 'Consortium For Medicare Health Plans Operations',
    legacy: true
  },
  OFFICE_OF_BURDEN_REDUCTION_AND_HEALTH_INFORMATICS: {
    name: 'Office Of Burden Reduction And Health Informatics',
    legacy: true
  },
  OFFICE_OF_SUPPORT_SERVICES_AND_OPERATIONS: {
    name: 'Office Of Support Services And Operations',
    legacy: true
  },
  OTHER: { name: 'Other', legacy: false }
};

export default cmsComponents;
