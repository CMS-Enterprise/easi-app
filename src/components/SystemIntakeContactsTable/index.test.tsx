import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import {
  render,
  screen,
  waitForElementToBeRemoved,
  within
} from '@testing-library/react';
import {
  SystemIntakeContactComponent,
  SystemIntakeContactFragment,
  SystemIntakeContactRole
} from 'gql/generated/graphql';
import i18next from 'i18next';
import {
  businessOwner,
  getSystemIntakeContactsQuery,
  productManager,
  requester,
  systemIntake
} from 'tests/mock/systemIntake';

import cmsComponents from 'constants/cmsComponents';

import SystemIntakeContactsTable from './index';

const additionalContact: SystemIntakeContactFragment = {
  __typename: 'SystemIntakeContact',
  systemIntakeId: systemIntake.id,
  id: '5783ca62-eafd-436e-aba1-1b74f37638be',
  userAccount: {
    __typename: 'UserAccount',
    id: '1811922d-e9a1-4727-903d-217c3cd4470a',
    username: 'USR1',
    commonName: 'User One',
    email: 'user.one@local.fake'
  },
  component: SystemIntakeContactComponent.OFFICE_OF_LEGISLATION,
  roles: [SystemIntakeContactRole.OTHER],
  isRequester: false
};

describe('SystemIntakeContactsTable', () => {
  it('matches the snapshot', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the contacts', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const allContacts: SystemIntakeContactFragment[] = [
      requester,
      businessOwner,
      productManager
    ];

    expect(screen.getAllByTestId('contact-row')).toHaveLength(
      allContacts.length
    );

    // Check that each contact renders in the table
    allContacts.forEach(contact => {
      expect(
        screen.getByRole('row', { name: `contact-${contact.id}` })
      ).toBeInTheDocument();
    });
  });

  it('displays requester icon for requester contact', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const requesterRow = screen.getByRole('row', {
      name: `contact-${requester.id}`
    });

    expect(requesterRow).toBeInTheDocument();

    const icon = within(requesterRow).getByRole('img', {
      name: 'Primary requester'
    });

    expect(icon).toBeInTheDocument();
  });

  it('does not display requester icon for non-requester contacts', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const nonRequesterRow = screen.getByRole('row', {
      name: `contact-${businessOwner.id}`
    });

    const icon = within(nonRequesterRow).queryByRole('img', {
      name: 'Primary requester'
    });

    expect(icon).not.toBeInTheDocument();
  });

  it('displays contact names and emails', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const requesterRow = screen.getByRole('row', {
      name: `contact-${requester.id}`
    });

    expect(
      within(requesterRow).getByText(requester.userAccount.commonName)
    ).toBeInTheDocument();
    expect(
      within(requesterRow).getByText(requester.userAccount.email)
    ).toBeInTheDocument();
  });

  it('displays component acronym', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const requesterRow = screen.getByRole('row', {
      name: `contact-${requester.id}`
    });

    const component = requester.component!;

    expect(
      within(requesterRow).getByText(cmsComponents[component].acronym!)
    ).toBeInTheDocument();
  });

  it('displays "None specified" for empty component', async () => {
    render(
      <MockedProvider
        mocks={[
          getSystemIntakeContactsQuery([
            {
              ...additionalContact,
              component: null
            }
          ])
        ]}
      >
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    expect(screen.getByText('None specified')).toBeInTheDocument();
  });

  it('displays translated role names', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const requesterRow = screen.getByRole('row', {
      name: `contact-${businessOwner.id}`
    });

    // Check that role is displayed (translated)
    expect(
      within(requesterRow).getByText(
        i18next.t<string>(
          `intake:contactDetails.systemIntakeContactRoles.${businessOwner.roles[0]}`
        )
      )
    ).toBeInTheDocument();
  });

  it('displays "None specified" for empty roles', async () => {
    render(
      <MockedProvider
        mocks={[
          getSystemIntakeContactsQuery([
            {
              ...additionalContact,
              roles: []
            }
          ])
        ]}
      >
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    expect(screen.getByText('None specified')).toBeInTheDocument();
  });

  it('renders edit and remove buttons for all contacts', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const businessOwnerRow = screen.getByRole('row', {
      name: `contact-${businessOwner.id}`
    });

    expect(
      within(businessOwnerRow).getByRole('button', { name: 'Edit' })
    ).toBeInTheDocument();
    expect(
      within(businessOwnerRow).getByRole('button', { name: 'Remove' })
    ).toBeInTheDocument();
  });

  it('disables remove button for requester contact', async () => {
    render(
      <MockedProvider mocks={[getSystemIntakeContactsQuery()]}>
        <SystemIntakeContactsTable systemIntakeId={systemIntake.id} />
      </MockedProvider>
    );

    await waitForElementToBeRemoved(() => screen.getByRole('progressbar'));

    const requesterRow = screen.getByRole('row', {
      name: `contact-${requester.id}`
    });

    const requesterRemoveButton = within(requesterRow).getByRole('button', {
      name: 'Remove'
    });

    expect(requesterRemoveButton).toBeDisabled();
  });
});
