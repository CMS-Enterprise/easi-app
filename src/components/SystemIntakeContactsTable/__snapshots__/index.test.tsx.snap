// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SystemIntakeContactsTable > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="system-intake-contacts-table usa-table-container--scrollable"
  >
    <table
      class="usa-table usa-table--borderless width-full"
      data-testid="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            aria-sort="none"
            class="border-bottom-2px"
            colspan="1"
            role="columnheader"
            scope="col"
            style="width: 150px;"
          >
            <button
              class="usa-button usa-button--unstyled width-full flex-justify margin-y-0"
              data-testid="button"
              style="cursor: pointer;"
              title="Toggle SortBy"
              type="button"
            >
              <span
                class="display-block margin-left-4 padding-left-05"
              >
                Name
              </span>
              <svg
                aria-hidden="true"
                class="usa-icon margin-left-05"
                data-testid="caret--sort"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            class="border-bottom-2px"
            colspan="1"
            role="columnheader"
            scope="col"
            style="width: 150px;"
          >
            <button
              class="usa-button usa-button--unstyled width-full flex-justify margin-y-0"
              data-testid="button"
              style="cursor: pointer;"
              title="Toggle SortBy"
              type="button"
            >
              Component
              <svg
                aria-hidden="true"
                class="usa-icon margin-left-05"
                data-testid="caret--sort"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            class="border-bottom-2px"
            colspan="1"
            role="columnheader"
            scope="col"
            style="width: 150px;"
          >
            <button
              class="usa-button usa-button--unstyled width-full flex-justify margin-y-0"
              data-testid="button"
              style="cursor: pointer;"
              title="Toggle SortBy"
              type="button"
            >
              Role(s)
              <svg
                aria-hidden="true"
                class="usa-icon margin-left-05"
                data-testid="caret--sort"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 5.83 15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
                />
              </svg>
            </button>
          </th>
          <th
            class="border-bottom-2px"
            colspan="1"
            role="columnheader"
            style="width: 150px;"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr>
          <td
            class="padding-left-4"
            colspan="4"
          >
            <span
              class="margin-left-05 display-flex flex-align-center text-italic"
            >
              <span
                aria-valuetext="Loading"
                class="easi-spinner easi-spinner--small margin-right-1"
                role="progressbar"
              />
              Loading contacts
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
